### PR TITLE
MSSQL 2014 OFFSET/FETCH syntax support

### LINKED ISSUE #5084: MSSQL 2012 OFFSET/FETCH syntax support
Looks like starting from version 2012, MSSQL  - supports normal limit/offset syntax.

With MSSQL 2012 this construction works:
```sql
SELECT product_name, list_price
FROM products
ORDER BY list_price, product_name 
OFFSET 20 ROWS 
FETCH NEXT 10 ROWS ONLY;
```

Is there a chance that we will get some option to support OFFSET/FETCH syntax with mssql?

### PR DESCRIPTION
**Pull request by ea-rus**
Added OFFSET/FETCH support for MSSQL 2014

### Description

Fixes: #5084

If version of MSSQL server is 11 or higher then limit offset is implementing by OFFSET and FETCH keywords as part of ORDER BY clause  instead subquery with row_number. 

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision 84d79d8acdea24c31de1b3bb4d7a63a2d3b9de8a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 84d79d8acdea24c31de1b3bb4d7a63a2d3b9de8a: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1695

File: lib/sqlalchemy/dialects/mssql/base.py [hunk 1]
```diff
@@ -2288,6 +2323,7 @@ class MSDialect(default.DefaultDialect):
     non_native_boolean_check_constraint = False
     supports_unicode_binds = True
     postfetch_lastrowid = True
+    _supports_offset_fetch = False
 
     server_version_info = ()
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/mssql/base.py]: I believe you also have to omit the usage of TOP in get_select_precolumns(), this is in my patch at https://github.com/sqlalchemy/sqlalchemy/issues/5084#issuecomment-572630940 which I'm assuming this is based off of

Comment (ea-rus) [reply lib/sqlalchemy/dialects/mssql/base.py]: Usage of TOP is possible without "ORDER BY". If we replace it with OFFSET/FETCH then sorting will be required. 

Moreover you can't use FETCH without OFFSET. This code returns an exception:

SELECT id from users
ORDER BY id
FETCH NEXT 1 ROWS ONLY 

Correct sql is:

SELECT id from users
ORDER BY id
OFFSET 0 ROWS
FETCH NEXT 1 ROWS ONLY 

So I think TOP should remain in cases where user choose simple limit, for example: 
db.Users.query.first()

Comment (zzzeek) [reply lib/sqlalchemy/dialects/mssql/base.py]: OK that makes sense, as long as we support all combinations.   I am hoping that all combinations are being exercised in https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/testing/suite/test_select.py#L150.   here's a pointless ASCII chart I made just now for no reason:

```
                       offset absent  |   offset simple integer  |  offset sql expression
                      +---------------+--------------------------+------------------------+
limit absent          |_______________|__________________________|________________________|
limit simple integer  |_______________|__________________________|________________________|
limit sql expression  |               |                          |                        |
                      +---------------+--------------------------+------------------------+
```

**Comment by zzzeek** [comment]
does the syntax support bound parameters?  if not, there's a flag you need to add to the "literals" so that they render inline rather than as parameters.   this will take advantage of a new feature in 1.4 that can render an inline literal after the statement is compiled.

`
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Invalid usage of the option NEXT in the FETCH statement. (153) (SQLExecDirectW)') [SQL: SELECT some_table.id, some_table.x, some_table.y  FROM some_table ORDER BY some_table.id  FETCH NEXT ? ROWS ONLY ]`

**Comment by ea-rus** [comment]
I think this error occurs because  FETCH was used without OFFSET. 
I can change function limit_clause in mssql/base.py and it will set OFFSET=0 if OFFSET is unset
This should solve these errors (in https://jenkins.sqlalchemy.org/job/sqlalchemy_gerrit/cext=cext,db=mssql-backendonly,pyv=py37/3698/):
- test.dialect.test_suite.LimitOffsetTest_mssql+pyodbc_14_0_3076_1.test_bound_limit
- test.dialect.test_suite.LimitOffsetTest_mssql+pyodbc_14_0_3076_1.test_expr_limit

As I understand this error, now DISTINCT works correctly with OFFSET/FETCH. By the way, I started using new syntax because DISTINCT does not work correct with subquery.
Should I do something with this error?
- test.sql.test_query.LimitTest_mssql+pyodbc_14_0_3076_1.test_select_distinct_offset

**Comment by zzzeek** [comment]
yes so the approach is, ,keep pushing changes here and flagging "sqla-tester" for review until all the tests pass. then we look at it and see what we have, if we like what it's doing or if there can be further refinements.

**Comment by zzzeek** [comment]
see if you have access to check "sqla-tester" in the list of reviewers in the upper right.  that triggers a new gerrit / CI run.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision 9345061440d58b6a36ce211cc0605339b1e7159c of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 9345061440d58b6a36ce211cc0605339b1e7159c added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1695

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision a9fda98d590ee16a9caa12ea50348c5fd1fbc825 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a9fda98d590ee16a9caa12ea50348c5fd1fbc825 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1695

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

care to take a shot at a changelog file?  

doc/build/changelog/unreleased_14/5084.rst

View this in Gerrit at https://gerrit.sqlalchemy.org/1695

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision a45b7f73090d2053e3a7020d4e3d7fabb0c5627d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a45b7f73090d2053e3a7020d4e3d7fabb0c5627d added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1695

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision a45b7f73090d2053e3a7020d4e3d7fabb0c5627d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a45b7f73090d2053e3a7020d4e3d7fabb0c5627d added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1695

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/1695

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1695 has been **merged**. Congratulations! :)

**Pull request closed without merging**